cpp/include/raft/cudart_utils.h

```diff
@@ -0,0 +1,224 @@
+//                      #call, __FILE__, __LINE__, cudaGetErrorString(status)); \
+//     }                                                                        \
+//   } while (0)
+
+/** helper method to get max usable shared mem per block parameter */
+inline int getSharedMemPerBlock() {
+  int devId;
+  CUDA_CHECK(cudaGetDevice(&devId));
+  int smemPerBlk;
+  CUDA_CHECK(cudaDeviceGetAttribute(&smemPerBlk,
+                                    cudaDevAttrMaxSharedMemoryPerBlock, devId));
+  return smemPerBlk;
+}
+/** helper method to get multi-processor count parameter */
+inline int getMultiProcessorCount() {
+  int devId;
+  CUDA_CHECK(cudaGetDevice(&devId));
+  int mpCount;
+  CUDA_CHECK(
+    cudaDeviceGetAttribute(&mpCount, cudaDevAttrMultiProcessorCount, devId));
+  return mpCount;
+}
+
+/**
+ * @brief Generic copy method for all kinds of transfers
+ * @tparam Type data type
+ * @param dst destination pointer
+ * @param src source pointer
+ * @param len lenth of the src/dst buffers in terms of number of elements
+ * @param stream cuda stream
+ */
+template <typename Type>
+void copy(Type* dst, const Type* src, size_t len, cudaStream_t stream) {
+  CUDA_CHECK(
+    cudaMemcpyAsync(dst, src, len * sizeof(Type), cudaMemcpyDefault, stream));
+}
+
+/**
+ * @defgroup Copy Copy methods
+ * These are here along with the generic 'copy' method in order to improve
+ * code readability using explicitly specified function names
+ * @{
+ */
+/** performs a host to device copy */
+template <typename Type>
+void updateDevice(Type* dPtr, const Type* hPtr, size_t len,
+                  cudaStream_t stream) {
+  copy(dPtr, hPtr, len, stream);
+}
+
+/** performs a device to host copy */
+template <typename Type>
+void updateHost(Type* hPtr, const Type* dPtr, size_t len, cudaStream_t stream) {
+  copy(hPtr, dPtr, len, stream);
+}
+
+template <typename Type>
+void copyAsync(Type* dPtr1, const Type* dPtr2, size_t len,
+               cudaStream_t stream) {
+  CUDA_CHECK(cudaMemcpyAsync(dPtr1, dPtr2, len * sizeof(Type),
+                             cudaMemcpyDeviceToDevice, stream));
+}
+/** @} */
+
+/**
+ * @defgroup Debug Utils for debugging device buffers
+ * @{
+ */
+template <class T, class OutStream>
+void printHostVector(const char* variableName, const T* hostMem,
+                     size_t componentsCount, OutStream& out) {
+  out << variableName << "=[";
+  for (size_t i = 0; i < componentsCount; ++i) {
+    if (i != 0) out << ",";
+    out << hostMem[i];
+  }
+  out << "];\n";
+}
+
+template <class T>
+void printHostVector(const char* variableName, const T* hostMem,
+                     size_t componentsCount) {
+  printHostVector(variableName, hostMem, componentsCount, std::cout);
+  std::cout.flush();
+}
+
+template <class T, class OutStream>
+void printDevVector(const char* variableName, const T* devMem,
+                    size_t componentsCount, OutStream& out) {
+  T* hostMem = new T[componentsCount];
```

Comment:
We also have a couple of helper functions that we should merge here as well (that's really a note to mystelf :) ). Sometimes we have duplicates with slightly different implementations like for printing we use :
`thrust::copy(dev_ptr, dev_ptr + n, std::ostream_iterator<T>(std::cout, " "));`

Reply:
Fair enough. Can you point me to those and I'll try to merge them the best way possible?

Reply:
This was mostly for visibility, it doesn't have to be in this PR unless you would like to reconcile now. 

cuGraph utils are here:
https://github.com/rapidsai/cugraph/blob/branch-0.14/cpp/src/utilities/graph_utils.cuh

Reply:
ok... device_utils and printing utilities can be further added in a separate PR. Makes sense. I have filed #9 for this one.